src/main/scala/com/fulcrumgenomics/bam/Pileup.scala

```diff
@@ -0,0 +1,209 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2017 Fulcrum Genomics LLC
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.fulcrumgenomics.bam
+
+import com.fulcrumgenomics.FgBioDef._
+import htsjdk.samtools.{SAMRecord, SAMSequenceDictionary}
+import htsjdk.samtools.util.SequenceUtil
+
+import scala.collection.mutable
+import scala.collection.mutable.ListBuffer
+
+/**
+  * Represents a pileup of reads/bases at a single genomic location.
+  */
+case class Pileup[A <: PileupEntry](refName: String, refIndex: Int, pos: Int, pile: Seq[A]) extends Iterable[A] {
+  override def iterator: Iterator[A] = pile.iterator
+
+  /**
+    * Returns the number of reads in the pileup.  This is subtley different from the number of entries in
+    * the pile as a given read may have two entries when the next base in the read is an insertion.
+    */
+  lazy val depth = pile.count {
+    case p: BaseEntry      => true
+    case d: DeletionEntry  => true
+    case i: InsertionEntry => i.offset == 0 /** Count insertions at the start of the read, but not elsewhere. */
```

Comment:
A leading insertion says nothing about the previous reference base though, so I don't think it should be counted here.  Lets chat about this since we have differing thoughts.